File: ArchitectureModel/src/main/java/com/kailin/architecture_model/util/LoggerUtil.java
package com.kailin.architecture_model.util;

import com.orhanobut.logger.AndroidLogAdapter;
import com.orhanobut.logger.Logger;

import java.util.concurrent.atomic.AtomicReference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class LoggerUtil {

    private static final AtomicReference<LoggerUtil> reference = new AtomicReference<>();

    public static LoggerUtil getInstance() {
        while (true) {
            LoggerUtil instance = reference.get();
            if (instance != null)
                return instance;

            instance = new LoggerUtil();
            if (reference.compareAndSet(null, instance))
                return instance;
        }
    }

    private LoggerUtil(){
        Logger.addLogAdapter(new AndroidLogAdapter());
    }

    public void log(int priority, @Nullable String tag, @Nullable String message, @Nullable Throwable throwable) {
        Logger.log(priority, tag, message, throwable);
    }

    public void d(@NonNull String message, @Nullable Object... args) {
        Logger.d(message, args);
    }

    public void d(@Nullable Object object) {
        Logger.d(object);
    }

    public void e(@NonNull String message, @Nullable Object... args) {
        Logger.e(null, message, args);
    }

    public void e(@Nullable Throwable throwable, @NonNull String message, @Nullable Object... args) {
        Logger.e(throwable, message, args);
    }

    public void i(@NonNull String message, @Nullable Object... args) {
        Logger.i(message, args);
    }

    public void v(@NonNull String message, @Nullable Object... args) {
        Logger.v(message, args);
    }

    public void w(@NonNull String message, @Nullable Object... args) {
        Logger.w(message, args);
    }

    public void wtf(@NonNull String message, @Nullable Object... args) {
        Logger.wtf(message, args);
    }

    public void json(@Nullable String json) {
        Logger.json(json);
    }

    public void xml(@Nullable String xml) {
        Logger.xml(xml);
    }
}

File: ArchitectureModel/src/main/java/com/kailin/architecture_model/util/LoggerUtil.kt
package com.kailin.architecture_model.util

import com.orhanobut.logger.AndroidLogAdapter
import com.orhanobut.logger.Logger

import java.util.concurrent.atomic.AtomicReference

class LoggerUtil private constructor() {

    init {
        Logger.addLogAdapter(AndroidLogAdapter())
    }

    fun log(priority: Int, tag: String?, message: String?, throwable: Throwable?) {
        Logger.log(priority, tag, message, throwable)
    }

    fun d(message: String, vararg args: Any) {
        Logger.d(message, *args)
    }

    fun d(`object`: Any?) {
        Logger.d(`object`)
    }

    fun e(message: String, vararg args: Any) {
        Logger.e(null, message, *args)
    }

    fun e(throwable: Throwable?, message: String, vararg args: Any) {
        Logger.e(throwable, message, *args)
    }

    fun i(message: String, vararg args: Any) {
        Logger.i(message, *args)
    }

    fun v(message: String, vararg args: Any) {
        Logger.v(message, *args)
    }

    fun w(message: String, vararg args: Any) {
        Logger.w(message, *args)
    }

    fun wtf(message: String, vararg args: Any) {
        Logger.wtf(message, *args)
    }

    fun json(json: String?) {
        Logger.json(json)
    }

    fun xml(xml: String?) {
        Logger.xml(xml)
    }

    companion object {

        private val reference = AtomicReference<LoggerUtil>()

        val instance: LoggerUtil
            get() {
                while (true) {
                    var instance: LoggerUtil? = reference.get()
                    if (instance != null)
                        return instance

                    instance = LoggerUtil()
                    if (reference.compareAndSet(null, instance))
                        return instance
                }
            }
    }
}
